### PR TITLE
LF-13019 LiFiDexAggregator to support Velodrome v2 [IVelodromeV2Pool v1.0.0,IVelodromeV2PoolCallee v1.0.0,IVelodromeV2PoolFactory v1.0.0,IVelodromeV2Router v1.0.0,LiFiDEXAggregator v1.8.0]

### PR DESCRIPTION
**Pull request by mirooon**
# Which Jira task belongs to this PR?
[LF-13019]

# Why did I implement it this way?

# Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] This pull request is as small as possible and only tackles one problem
- [ ] I have added tests that cover the functionality / test the bug
- [ ] For new facets: I have checked all points from this list: https://www.notion.so/lifi/New-Facet-Contract-Checklist-157f0ff14ac78095a2b8f999d655622e
- [ ] I have updated any required documentation

# Checklist for reviewer (DO NOT DEPLOY and contracts BEFORE CHECKING THIS!!!)

- [ ] I have checked that any arbitrary calls to external contracts are validated and or restricted
- [ ] I have checked that any privileged calls (i.e. storage modifications) are validated and or restricted
- [ ] I have ensured that any new contracts have had AT A MINIMUM 1 preliminary audit conducted on <date> by <company/auditor>

[LF-13019]: https://lifi.atlassian.net/browse/LF-13019?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: test/solidity/utils/TestBase.sol [hunk 1]
```diff
@@ -192,6 +192,17 @@ abstract contract TestBase is Test, DiamondTest, ILiFi {
...
+    address internal constant ADDRESS_UNISWAP_OPTIMISM =
+        0x4A7b5Da61326A6379179b40d00F57E5bbDC962c2;
+    address internal constant ADDRESS_USDC_OPTIMISM =
+        0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85;
+    address internal constant ADDRESS_USDT_OPTIMISM =
+        0x94b008aA00579c1307B0EF2c499aD98a8ce58e58;
+    address internal constant ADDRESS_DAI_OPTIMISM =
+        0xDA10009cBd5D07dd0CeCc66161FC93D7c9000da1;
+    address internal constant ADDRESS_WRAPPED_NATIVE_OPTIMISM =
+        0x4200000000000000000000000000000000000006;
```
[the remark this answers is not in the document]
Comment (mirooon) [reply test/solidity/utils/TestBase.sol:205]: For ADDRESS_UNISWAP_OPTIMISM its for uniswap router not uniswap token
For ADDRESS_USDC_OPTIMISM current address is for USDC and updated is for USDC.e.

File: src/Periphery/LiFiDEXAggregator.sol [hunk 2]
```diff
@@ -184,7 +189,9 @@ contract LiFiDEXAggregator is WithdrawablePeriphery {
                     if (step == 0) realAmountIn = usedAmount;
                 } else if (commandCode == 4) processOnePool(stream);
                 else if (commandCode == 5) processInsideBento(stream);
-                else if (commandCode == 6) applyPermit(tokenIn, stream);
+                else if (commandCode == 6)
```
Comment (0xDEnYO) [inline src/Periphery/LiFiDEXAggregator.sol:192]: Could you please explain the rationale behind this code and why it had to be added (and maybe add it as a comment)?

Will Velodrome will only support tokens that support EIP2612?

File: src/Periphery/LiFiDEXAggregator.sol [hunk 3]
```diff
@@ -750,6 +765,45 @@ contract LiFiDEXAggregator is WithdrawablePeriphery {
...
+        address from,
+        address tokenIn,
+        uint256 amountIn
+    ) private {
+        address pool = stream.readAddress();
+        uint8 direction = stream.readUint8();
+        address to = stream.readAddress();
+        stream.readUint24(); // pool fee in 1/1_000_000
+        stream.readUint8() > 0;
+        bool callback = stream.readUint8() > 0; // if true then run callback after swap with tokenIn as flashloan data. Will revert if contract (to) does not impelent IVelodromeV2PoolCallee
```
Comment (0xDEnYO) [inline src/Periphery/LiFiDEXAggregator.sol:780]: typo in comment "impelent"

File: src/Periphery/LiFiDEXAggregator.sol [hunk 4]
```diff
@@ -750,6 +765,45 @@ contract LiFiDEXAggregator is WithdrawablePeriphery {
...
+        // we don't need 'fee' and 'stable' flags since the pool handles that internally
```
Comment (0xDEnYO) [inline src/Periphery/LiFiDEXAggregator.sol:781]: would propose the following update: 

```solidity
// we don't need 'pool fee' and 'stable' flag since the pool handles that internally but (must) read them anyway to forward to pointer of the stream
stream.readUint24(); // read pool fee (in 1/1_000_000)
stream.readUint8(); // read stable flag 
```


note: I removed the `> 0` as this is useless and a waste of gas

File: test/solidity/Periphery/LiFiDEXAggregator.t.sol [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+        uint24 fee;
+        uint8 direction;
+        bool callback;
+    }
+
+    /**
+     * @dev Helper function to test a VelodromeV2 swap.
+     * Uses a struct to group parameters and reduce stack depth.
+     */
+    function _testSwap(SwapTestParams memory params) internal {
```
Comment (0xDEnYO) [inline test/solidity/Periphery/LiFiDEXAggregator.t.sol:212]: very nicely coded (and commented 😀)  !!!

File: test/solidity/Periphery/LiFiDEXAggregator.t.sol [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+        // fund dex aggregator contract so that the contract holds USDC
+        deal(address(USDC_TOKEN), address(liFiDEXAggregator), 100_000 * 1e6);
+
+        vm.startPrank(USER_SENDER);
+        _testSwap(
+            SwapTestParams({
+                from: address(liFiDEXAggregator),
+                to: address(USER_SENDER),
+                tokenIn: address(USDC_TOKEN),
+                amountIn: USDC_TOKEN.balanceOf(address(liFiDEXAggregator)) - 1, // has to be current dex aggregator balance - 1
```
Comment (0xDEnYO) [inline test/solidity/Periphery/LiFiDEXAggregator.t.sol:165]: why has to be -1 ?

File: test/solidity/Periphery/LiFiDEXAggregator.t.sol [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+                stable: true, // - NOT USED!
+                fee: 500, // - NOT USED!
+                direction: 0,
+                callback: true
+            })
+        );
+        vm.stopPrank();
+    }
+
+    struct SwapTestParams {
```
Comment (0xDEnYO) [inline test/solidity/Periphery/LiFiDEXAggregator.t.sol:196]: suggest to move to the top of the contract (before it's being used)

File: test/solidity/Periphery/LiFiDEXAggregator.t.sol [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+        emit HookCalled(sender, amount0, amount1, data);
+    }
+}
+
+contract LiFiDexAggregator is TestBase {
+    IVelodromeV2Router internal constant VELODROME_V2_ROUTER =
+        IVelodromeV2Router(0xa062aE8A9c5e11aaA026fc2670B0D65cCc8B2858); // optimism router
+    address internal constant VELODROME_V2_FACTORY_REGISTRY =
+        0xF1046053aa5682b4F9a81b5481394DA16BE5FF5a;
+    IERC20 internal constant USDC_TOKEN =
```
Comment (0xDEnYO) [inline test/solidity/Periphery/LiFiDEXAggregator.t.sol:33]: you dont need this variable since you already defined it in TestBase as `ADDRESS_USDC_OPTIMISM`

File: test/solidity/Periphery/LiFiDEXAggregator.t.sol (hunk 6 above)
Comment (mirooon) [reply test/solidity/Periphery/LiFiDEXAggregator.t.sol]: good question

its because undrain safe protection that is added during processing LDA erc20 tokens.

https://github.com/lifinance/contracts/blob/main/src/Periphery/LiFiDEXAggregator.sol#L254-L262

don't know the exact use case for it but I assume that it's additional safety layer against some edge cases when balance is fully drained. I've updated the comment to more precise

File: test/solidity/Periphery/LiFiDEXAggregator.t.sol (hunk 7 above)
Comment (mirooon) [reply test/solidity/Periphery/LiFiDEXAggregator.t.sol:196]: updated

File: test/solidity/Periphery/LiFiDEXAggregator.t.sol (hunk 5 above)
Comment (mirooon) [reply test/solidity/Periphery/LiFiDEXAggregator.t.sol:212]: thank you!

File: src/Periphery/LiFiDEXAggregator.sol (hunk 2 above)
Comment (mirooon) [reply src/Periphery/LiFiDEXAggregator.sol]: I didn't add it. It was just formatting. 

Regarding the permit functionality for velodrome the dex itself doesn't support EIP2612 tokens. Only LP tokens can have gasless approvals.

File: src/Periphery/LiFiDEXAggregator.sol [hunk 9]
```diff
@@ -388,29 +400,29 @@ contract LiFiDEXAggregator is WithdrawablePeriphery {
...
-                    IERC20(tokenIn).balanceOf(address(bentoBox)) +
-                    bentoBox.strategyData(tokenIn).balance -
-                    bentoBox.totals(tokenIn).elastic;
+                    IERC20(tokenIn).balanceOf(address(BENTO_BOX)) +
+                    BENTO_BOX.strategyData(tokenIn).balance -
+                    BENTO_BOX.totals(tokenIn).elastic;
             }
-            bentoBox.deposit(tokenIn, address(bentoBox), to, amountIn, 0);
+            BENTO_BOX.deposit(tokenIn, address(BENTO_BOX), to, amountIn, 0);
         } else {
```
[the remark this answers is not in the document]
Comment (mirooon) [reply src/Periphery/LiFiDEXAggregator.sol:424]: good point

File: src/Interfaces/IVelodromeV2Pool.sol [hunk 10]
```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.17;
+
+/// @title Interface for VelodromeV2 pool
+/// @author LI.FI (https://li.fi)
+/// @custom:version 1.0.0
+interface IVelodromeV2Pool {
+    /// @notice This low-level function should be called from a contract which performs important safety checks
```
[the remark this answers is not in the document]
Comment (mirooon) [reply src/Interfaces/IVelodromeV2Pool.sol:8]: Covered things:

1. Reentrancy - LDA has custom lock() modifier 
2. we use SafeERC20 for token transfers to make sure we revert on failure
3. before executing swap the contract call getAmountOut to calculate expected output (including fee).
4. flashloan flag to know if the callback should be triggered
5. in processRouteInternal after executing all swaps we check minAmountOut for recipient and checks if the sender's final balance is not less than the initial balance.
6. also it's worth to mention that Velodrome V2 pools and router are not upgradable and we can rely on functions that we interact with (getAmountOut and swap)

Do you see anything more that we should add?

[the remark this answers is not in the document]
Comment (0xDEnYO) [reply src/Interfaces/IVelodromeV2Pool.sol:8]: Sounds good. No I do not see any additional things. But I did not review the Velodrome protocol/contracts in detail.
Maybe add this as a comment for the auditor to verify that this is safe, ok?

File: src/Periphery/LiFiDEXAggregator.sol (hunk 2 above)
Comment (0xDEnYO) [reply src/Periphery/LiFiDEXAggregator.sol]: ahhh...lol...OK, did not notice that, sorry.

File: test/solidity/Periphery/LiFiDEXAggregator.t.sol (hunk 6 above)
Comment (0xDEnYO) [reply test/solidity/Periphery/LiFiDEXAggregator.t.sol]: FYI (chatGPT's take to this):

image

File: src/Periphery/LiFiDEXAggregator.sol [hunk 11]
```diff
@@ -69,7 +79,10 @@ contract LiFiDEXAggregator is WithdrawablePeriphery {
         address[] memory priviledgedUserList,
         address _owner
     ) WithdrawablePeriphery(_owner) {
-        bentoBox = IBentoBoxMinimal(_bentoBox);
+        if (_bentoBox == address(0) || _owner == address(0)) {
```
Comment (0xDEnYO) [inline src/Periphery/LiFiDEXAggregator.sol:82]: this is incorrect. We currently deploy the LDA with _bentoBox == address(0) so this check (and the respective test) needs to be removed

**Review by 0xDEnYO — CHANGES_REQUESTED** [review]

File: test/solidity/Periphery/LiFiDEXAggregator.t.sol [hunk 12]
```diff
@@ -72,15 +74,44 @@ contract LiFiDexAggregatorTest is TestBase {
...
     }
 
+    function test_ContractIsSetUpCorrectly() public {
+        assertEq(address(liFiDEXAggregator.BENTO_BOX()), address(0xCAFE));
+        assertEq(liFiDEXAggregator.priviledgedUsers(address(0xABC)), true);
+        assertEq(liFiDEXAggregator.priviledgedUsers(address(0xEBC)), true);
+        assertEq(liFiDEXAggregator.owner(), USER_DIAMOND_OWNER);
+    }
+
+    function testRevert_FailsIfBentoBoxIsZeroAddress() public {
```
Comment (0xDEnYO) [inline test/solidity/Periphery/LiFiDEXAggregator.t.sol:95]: pls remove

File: src/Periphery/LiFiDEXAggregator.sol (hunk 11 above)
Comment (mirooon) [reply src/Periphery/LiFiDEXAggregator.sol:82]: true! updated.

File: src/Interfaces/IVelodromeV2Pool.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (mirooon) [reply src/Interfaces/IVelodromeV2Pool.sol:8]: added

**Review by 0xDEnYO — APPROVED** [review]

File: test/solidity/Periphery/LiFiDEXAggregator.t.sol [hunk 13]
```diff
@@ -0,0 +1,898 @@
...
+    );
+    event HookCalled(
+        address sender,
+        uint256 amount0,
+        uint256 amount1,
+        bytes data
+    );
+
+    error WrongPoolReserves();
+
```
[the remark this answers is not in the document]
Comment (mirooon) [reply test/solidity/Periphery/LiFiDEXAggregator.t.sol:63]: its good, we can have duplicated errors inside test files

File: src/Periphery/LiFiDEXAggregator.sol [hunk 14]
```diff
@@ -750,6 +759,77 @@
...
+            else if (from == msg.sender)
+                IERC20(tokenIn).safeTransferFrom(msg.sender, pool, amountIn);
+        }
+
+        // calculate the expected output amount using the pool's getAmountOut function
+        uint256 amountOut = IVelodromeV2Pool(pool).getAmountOut(
+            amountIn,
+            tokenIn
+        );
+
```
[the remark this answers is not in the document]
Comment (mirooon) [reply src/Periphery/LiFiDEXAggregator.sol:806]: reserves won't be updated before swap function execution

File: src/Periphery/LiFiDEXAggregator.sol [hunk 15]
```diff
@@ -750,6 +759,77 @@ contract LiFiDEXAggregator is WithdrawablePeriphery {
...
+        // set the appropriate output amount based on which token is being swapped
+        // determine output amounts based on direction
+        uint256 amount0Out = direction == 0 ? 0 : amountOut;
+        uint256 amount1Out = direction == 0 ? amountOut : 0;
+
+        // 'swap' function from IVelodromeV2Pool should be called from a contract which performs important safety checks.
```
[the remark this answers is not in the document]
Comment (mirooon) [reply src/Periphery/LiFiDEXAggregator.sol:812]: good point

File: src/Periphery/LiFiDEXAggregator.sol [hunk 16]
```diff
@@ -750,6 +759,74 @@ contract LiFiDEXAggregator is WithdrawablePeriphery {
...
+        // ATTENTION FOR CALLBACKS / HOOKS:
+        // - recipient contracts should validate that msg.sender is the Velodrome pool contract who is calling the hook
+        // - recipient contracts must not manipulate their own tokenOut balance (as this may bypass/invalidate the built-in slippage protection)
+        // - @developers: never trust balance-based slippage protection for callback recipients
+        // - @integrators: do not use slippage guarantees when recipient is a contract with side-effects
+        IVelodromeV2Pool(pool).swap(
+            amount0Out,
+            amount1Out,
+            to,
+            callback ? abi.encode(tokenIn) : new bytes(0)
```
Comment (ezynda3) [inline src/Periphery/LiFiDEXAggregator.sol:827]: I think using `bytes("")` is more gas efficient as `new` allocates memory

File: src/Periphery/LiFiDEXAggregator.sol [hunk 17]
```diff
@@ -750,6 +759,74 @@ contract LiFiDEXAggregator is WithdrawablePeriphery {
...
+        address to = stream.readAddress();
+        if (pool == address(0) || to == address(0)) revert InvalidCallData();
+        bool callback = stream.readUint8() == 1; // if true then run callback after swap with tokenIn as flashloan data. Will revert if contract (to) does not implement IVelodromeV2PoolCallee
+
+        if (from == INTERNAL_INPUT_SOURCE) {
+            // processOnePool case
+            (uint256 reserve0, uint256 reserve1, ) = IVelodromeV2Pool(pool)
+                .getReserves();
+            if (reserve0 == 0 || reserve1 == 0) revert WrongPoolReserves();
+            uint256 reserveIn = direction == 1 ? reserve0 : reserve1;
```
[the remark this answers is not in the document]
Comment (mirooon) [reply src/Periphery/LiFiDEXAggregator.sol]: Good point!
Added DIRECTION_TOKEN1_TO_TOKEN0. DIRECTION_TOKEN0_TO_TOKEN1 currently would be not needed

File: src/Periphery/LiFiDEXAggregator.sol (hunk 16 above)
Comment (mirooon) [reply src/Periphery/LiFiDEXAggregator.sol]: updated

**Review by ezynda3 — APPROVED** [review]

File: audit/auditLog.json [hunk 18]
```diff
@@ -209,6 +216,13 @@
       "auditorGitHandle": "sujithsomraaj",
       "auditReportPath": "./audit/reports/2025.04.15_Permit2Proxy(v1.0.3).pdf",
       "auditCommitHash": "5b212bd302a7ec0c027c51753d6d5a9937f6769e"
+    },
+    "audit20250422": {
+      "auditCompletedOn": "22.04.2025",
+      "auditedBy": "Sujith Somraaj (individual security researcher)",
+      "auditorGitHandle": "sujithsomraaj",
+      "auditReportPath": "./audit/reports/2025.04.22_LiFiDEXAggregator(v1.8.0).pdf",
```
Comment (ezynda3) [inline audit/auditLog.json:224]: Should be `./audit/reports/2025.04.22_LiFiDexAggregator(v1.8.0).pdf`

**Review by ezynda3 — CHANGES_REQUESTED** [review]

File: audit/auditLog.json (hunk 18 above)
Comment (mirooon) [reply audit/auditLog.json]: updated!

**Review by ezynda3 — APPROVED** [review]
